src/Propulsion/Streams.fs

```diff
@@ -155,8 +155,9 @@ module Buffer =
         member x.EventsCount = if x.IsEmpty then 0 else x.queue |> Seq.sumBy Array.length
 
         member x.HeadSpan = x.queue[0]
+        member x.QueuePos = x.HeadSpan[0].Index
         member x.IsMalformed = not x.IsEmpty && WritePosMalformed = x.write
-        member x.HasGap = match x.write with WritePosUnknown -> false | w -> w <> x.HeadSpan[0].Index
+        member x.HasGap = match x.write with WritePosUnknown -> not x.IsEmpty && x.QueuePos <> 0L | w -> w <> x.QueuePos
```

Comment:
Attn @nordfjord not sure if you have any variant of this DNA anywhere
Bottom line is that there is no useful way to check/detect/handle out of order delivery and/or gap detection without requiring all processing of all streams to start from zero
It's conceivable one might build a mechanism where things can be delivered speculatively, allowing the handler to say "the outcome of handling event 15 and 16 as presented before we knew the write position is: we are already at position 20 and you can start dropping stuff before that with confidence" if such a mechanism was introduced, then detecting and/or waiting for correction of gaps could be implemented, but any such thing would need to be rigorously assessed wrt better ways of handling the problem as a whole